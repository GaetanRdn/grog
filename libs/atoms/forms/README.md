# atoms-forms

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test atoms-forms` to execute the unit tests.
