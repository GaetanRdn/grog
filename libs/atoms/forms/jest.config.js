module.exports = {
  displayName: 'atoms-forms',
  preset: '../../../jest.preset.js',
  setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],
  globals: {
    'ts-jest': {
      tsconfig: '<rootDir>/tsconfig.spec.json',
      stringifyContentPathRegex: '\\.(html|svg)$',
    },
  },
  collectCoverage: true,
  coverageDirectory: '../../../coverage/libs/atoms/forms',
  transform: {
    '^.+\\.(ts|js|html)$': 'jest-preset-angular',
  },
  snapshotSerializers: [
    'jest-preset-angular/build/serializers/no-ng-attributes',
    'jest-preset-angular/build/serializers/ng-snapshot',
    'jest-preset-angular/build/serializers/html-comment',
  ],
};
