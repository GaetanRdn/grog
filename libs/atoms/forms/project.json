{
  "projectType": "library",
  "root": "libs/atoms/forms",
  "sourceRoot": "libs/atoms/forms/src",
  "prefix": "gro",
  "targets": {
    "build": {
      "executor": "@nrwl/angular:ng-packagr-lite",
      "outputs": ["dist/libs/atoms/forms"],
      "options": {
        "project": "libs/atoms/forms/ng-package.json"
      },
      "configurations": {
        "production": {
          "tsConfig": "libs/atoms/forms/tsconfig.lib.prod.json"
        },
        "development": {
          "tsConfig": "libs/atoms/forms/tsconfig.lib.json"
        }
      },
      "defaultConfiguration": "production"
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/libs/atoms/forms"],
      "options": {
        "jestConfig": "libs/atoms/forms/jest.config.js",
        "passWithNoTests": true
      }
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "options": {
        "lintFilePatterns": [
          "libs/atoms/forms/src/**/*.ts",
          "libs/atoms/forms/src/**/*.html"
        ]
      }
    },
    "storybook": {
      "executor": "@nrwl/storybook:storybook",
      "options": {
        "uiFramework": "@storybook/angular",
        "port": 4400,
        "config": {
          "configFolder": "libs/atoms/forms/.storybook"
        },
        "projectBuildConfig": "atoms-forms"
      },
      "configurations": {
        "ci": {
          "quiet": true
        }
      }
    },
    "build-storybook": {
      "executor": "@nrwl/storybook:build",
      "outputs": ["{options.outputPath}"],
      "options": {
        "uiFramework": "@storybook/angular",
        "outputPath": "dist/storybook/atoms-forms",
        "config": {
          "configFolder": "libs/atoms/forms/.storybook"
        },
        "projectBuildConfig": "atoms-forms"
      },
      "configurations": {
        "ci": {
          "quiet": true
        }
      }
    }
  },
  "tags": [],
  "implicitDependencies": ["styles", "tests", "types"]
}
