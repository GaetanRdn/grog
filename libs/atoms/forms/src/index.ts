export * from './lib/input/input.directive';
export * from './lib/autocomplete/autocomplete.component';
export * from './lib/label/label.component';
export * from './lib/checkbox/checkbox.component';
export * from './lib/select/select.component';
