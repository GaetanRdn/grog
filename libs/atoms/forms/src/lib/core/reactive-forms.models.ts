export type OnChangeFn<ValueType> = (value: ValueType | null) => void;
export type OnTouchedFn = () => void;
