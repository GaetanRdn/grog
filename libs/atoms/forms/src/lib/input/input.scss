@use '../../../../../styles/src/lib/utils';

@mixin gro-input($config) {
  $typography: map-get(map-get($config, typography), input);
  $colors: map-get($config, colors);
  $font-size: map-get($typography, font-size);

  .gro-input {
    @include utils.set-color-values($config, primary);
    font-size: $font-size;
    font-family: map-get($typography, font-family);
    font-weight: map-get($typography, font-weight);
    border-radius: 0.25 * $font-size;
    outline: none;
    height: 2 * $font-size;

    &:not(.gro-disabled, .gro-readonly) {
      @include utils.set-color-values($config, primary);

      &:hover,
      &:focus {
        @include utils.set-color-values($config, primary-dark);
      }
    }

    &.gro-disabled,
    &.gro-readonly {
      @include utils.set-color-values($config, disabled);
      border-width: 1px;
      border-style: solid;
    }
  }
}
