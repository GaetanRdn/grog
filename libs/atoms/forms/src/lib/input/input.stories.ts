import { Meta, Story } from '@storybook/angular';
import { InputDirective } from './input.directive';

export default {
  title: 'atoms/forms/input',
  argTypes: {
    value: { control: { type: 'text' } },
    valueChange: { action: 'valueChange' },
  },
  component: InputDirective,
  parameters: {
    jest: ['input.directive.spec.ts'],
  },
} as Meta<InputDirective<unknown>>;

const BasicTemplate: Story<InputDirective<unknown>> = (args) => ({
  props: {
    ...args,
  },
  template: `<input groInput [value]='value' (valueChange)='valueChange($event)' [disabled]='disabled' [readonly]='readonly'/>`,
});

export const Default = BasicTemplate.bind({});
Default.args = {
  value: '',
  disabled: false,
  readonly: false,
};
