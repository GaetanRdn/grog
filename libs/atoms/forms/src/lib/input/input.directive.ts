import { CommonModule } from '@angular/common';
import {
  Directive,
  ElementRef,
  EventEmitter,
  HostListener,
  Input,
  NgModule,
  OnChanges,
  Output,
} from '@angular/core';
import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
import { AutoUnsubscribe, CoerceBoolean } from '@grorg/decorators';
import { BooleanInput, NgChanges, Nullable } from '@grorg/types';
import { OnChangeFn, OnTouchedFn } from '../core/reactive-forms.models';

@Directive({
  selector: 'input[groInput]',
  host: {
    class: 'gro-input',
    '[class.gro-focused]': 'focused || null',
    '[class.gro-readonly]': 'readonly || null',
    '[class.gro-disabled]': 'disabled || null',
    '[attr.value]': 'value',
    '[attr.readonly]': 'readonly || null',
    '[disabled]': 'disabled || null',
  },
  providers: [
    { provide: NG_VALUE_ACCESSOR, useExisting: InputDirective, multi: true },
  ],
})
@AutoUnsubscribe()
export class InputDirective implements ControlValueAccessor, OnChanges {
  static ngAcceptInputType_disabled: BooleanInput;

  @Input()
  public value = '';

  @Input()
  @CoerceBoolean()
  public readonly = false;

  @Output()
  public readonly valueChange: EventEmitter<InputDirective['value']> =
    new EventEmitter<InputDirective['value']>();

  @Input()
  @CoerceBoolean()
  public disabled = false;

  private _focused = false;

  get focused(): boolean {
    return this._focused;
  }

  constructor(private readonly _elementRef: ElementRef<HTMLInputElement>) {}

  public ngOnChanges(changes: NgChanges<InputDirective>): void {
    if (changes.value) {
      this._elementRef.nativeElement.value = this.value;
    }
  }

  public writeValue(value: Nullable<InputDirective['value']>): void {
    this.value = value ?? '';
  }

  public registerOnChange(
    fn: OnChangeFn<Nullable<InputDirective['value']>>
  ): void {
    this.onChange = fn;
  }

  public registerOnTouched(fn: OnTouchedFn): void {
    this.onTouched = fn;
  }

  public setDisabledState(disabled: boolean): void {
    this.disabled = disabled;
  }

  @HostListener('input', ['$event.target.value'])
  public onInput(value: string): void {
    if (!this.readonly && !this.disabled) {
      this.value = value;
      this.valueChange.emit(value);
      this.onChange(value);
      this.onTouched();
    }
  }

  @HostListener('focus', ['true'])
  @HostListener('blur', ['false'])
  public onToggleFocus(focused: boolean): void {
    if (!this.readonly && !this.disabled) {
      this._focused = focused;
      this.onTouched();
    }
  }

  public onChange: OnChangeFn<Nullable<InputDirective['value']>> = (): void => {
    // default if no ngControl
  };

  public onTouched: OnTouchedFn = (): void => {
    // default if no ngControl
  };
}

@NgModule({
  declarations: [InputDirective],
  imports: [CommonModule],
  exports: [InputDirective],
})
export class InputModule {}
