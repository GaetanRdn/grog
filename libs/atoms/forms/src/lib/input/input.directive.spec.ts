import { Component, DebugElement, OnDestroy } from '@angular/core';
import { TestBed } from '@angular/core/testing';
import { FormControl, ReactiveFormsModule } from '@angular/forms';
import { TemplateLookup } from '@grorg/tests';
import { InputDirective } from './input.directive';

describe('InputDirective', () => {
  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [
        InputDirective,
        BasicComponent,
        ReadonlyComponent,
        DisabledComponent,
        FormControlComponent,
      ],
      imports: [ReactiveFormsModule],
    });
  });

  describe('Basic input', () => {
    let templateLookup: TemplateLookup<BasicComponent>;

    beforeEach(() => {
      templateLookup = new TemplateLookup(
        TestBed.createComponent(BasicComponent)
      );
    });

    test('Check default state', () => {
      // WHEN
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test.each(['value1', 'value2'])('should set value %s', (value: string) => {
      // GIVEN
      templateLookup.hostComponent.value = value;

      // WHEN
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(templateLookup.query<HTMLInputElement>('input').value).toEqual(
        value
      );
    });

    test.each(['value1', 'value2'])(
      'should update parent value on change, expect %s',
      (value: string) => {
        // GIVEN
        templateLookup.detectChanges();

        // WHEN
        triggerInputValue(templateLookup, value);
        templateLookup.detectChanges();

        // THEN
        expect(templateLookup.firstChildElement).toMatchSnapshot();
        expect(templateLookup.hostComponent.value).toEqual(value);
      }
    );

    test('check class on focus', () => {
      // GIVEN
      templateLookup.detectChanges();

      // WHEN
      const input: DebugElement = templateLookup.get('input');
      input.triggerEventHandler('focus', input.nativeElement);
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('when focused then check class on blur', () => {
      // GIVEN
      const input: DebugElement = templateLookup.get('input');
      input.triggerEventHandler('focus', input.nativeElement);
      templateLookup.detectChanges();

      // WHEN
      input.triggerEventHandler('blur', input.nativeElement);
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('check OnDestroy', () => {
      // GIVEN
      const input: InputDirective = templateLookup
        .get('input')
        .injector.get(InputDirective);
      jest.spyOn(input.valueChange, 'unsubscribe');

      // WHEN
      (input as unknown as OnDestroy).ngOnDestroy();

      // THEN
      expect(input.valueChange.unsubscribe).toBeCalled();
    });
  });

  describe('Readonly input', () => {
    let templateLookup: TemplateLookup<ReadonlyComponent>;

    beforeEach(() => {
      templateLookup = new TemplateLookup(
        TestBed.createComponent(ReadonlyComponent)
      );

      templateLookup.detectChanges();
    });

    test('do not emit change', () => {
      // WHEN
      triggerInputValue(templateLookup, 'value');
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(templateLookup.hostComponent.value).toBeNull();
    });

    test('when focus then check dont have focused class', () => {
      // GIVEN
      templateLookup.detectChanges();

      // WHEN
      const input: DebugElement = templateLookup.get('input');
      input.triggerEventHandler('focus', input.nativeElement);
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });
  });

  describe('Disabled input', () => {
    let templateLookup: TemplateLookup<DisabledComponent>;

    beforeEach(() => {
      templateLookup = new TemplateLookup(
        TestBed.createComponent(DisabledComponent)
      );

      templateLookup.detectChanges();
    });

    test('do not emit change', () => {
      // WHEN
      triggerInputValue(templateLookup, 'value');
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(templateLookup.hostComponent.value).toBeNull();
    });

    test('when focus then check dont have focused class', () => {
      // GIVEN
      templateLookup.detectChanges();

      // WHEN
      const input: DebugElement = templateLookup.get('input');
      input.triggerEventHandler('focus', input.nativeElement);
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });
  });

  describe('ReactiveForm input', () => {
    let templateLookup: TemplateLookup<FormControlComponent>;

    beforeEach(() => {
      templateLookup = new TemplateLookup(
        TestBed.createComponent(FormControlComponent)
      );

      templateLookup.detectChanges();
    });

    test('should create with formControl', () => {
      // WHEN
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(
        templateLookup.query<HTMLInputElement>(InputDirective).value
      ).toEqual('');
    });

    test('should set value from formControl', () => {
      // GIVEN
      templateLookup.hostComponent.control.setValue('test');

      // WHEN
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(
        templateLookup.query<HTMLInputElement>(InputDirective).value
      ).toEqual('test');
    });

    test('should set value to formControl', async () => {
      // GIVEN
      templateLookup.detectChanges();

      // WHEN
      triggerInputValue(templateLookup, 'toto');
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(templateLookup.hostComponent.control.value).toEqual('toto');
    });

    test('check class on focus', () => {
      // GIVEN
      templateLookup.detectChanges();

      // WHEN
      const input: DebugElement = templateLookup.get('input');
      input.triggerEventHandler('focus', input.nativeElement);
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('when control is disabled then check', () => {
      // GIVE
      templateLookup.hostComponent.control.disable();

      // WHEN
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });
  });
});

const triggerInputValue = (
  templateLookup: TemplateLookup<unknown>,
  value: string
): void => {
  const input: DebugElement = templateLookup.get('input');
  (input.nativeElement as HTMLInputElement).value = value;
  input.triggerEventHandler('input', { target: input.nativeElement });
};

@Component({
  template: `<input groInput [(value)]="value" />`,
})
class BasicComponent {
  public value!: string;
}

@Component({
  template: `<input groInput readonly [(value)]="value" />`,
})
class ReadonlyComponent {
  public value = null;
}

@Component({
  template: `<input groInput disabled [(value)]="value" />`,
})
class DisabledComponent {
  public value = null;
}

@Component({
  template: `<input groInput [formControl]="control" />`,
})
class FormControlComponent {
  public control: FormControl = new FormControl();
}
