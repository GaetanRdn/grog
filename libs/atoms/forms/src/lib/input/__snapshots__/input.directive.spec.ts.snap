// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputDirective Basic input Check default state 1`] = `
<input
  class="gro-input"
  groinput=""
/>
`;

exports[`InputDirective Basic input check class on focus 1`] = `
<input
  class="gro-input gro-focused"
  groinput=""
/>
`;

exports[`InputDirective Basic input should set value value1 1`] = `
<input
  class="gro-input"
  groinput=""
  value="value1"
/>
`;

exports[`InputDirective Basic input should set value value2 1`] = `
<input
  class="gro-input"
  groinput=""
  value="value2"
/>
`;

exports[`InputDirective Basic input should update parent value on change, expect value1 1`] = `
<input
  class="gro-input"
  groinput=""
  value="value1"
/>
`;

exports[`InputDirective Basic input should update parent value on change, expect value2 1`] = `
<input
  class="gro-input"
  groinput=""
  value="value2"
/>
`;

exports[`InputDirective Basic input when focused then check class on blur 1`] = `
<input
  class="gro-input"
  groinput=""
/>
`;

exports[`InputDirective Disabled input do not emit change 1`] = `
<input
  class="gro-input gro-disabled"
  disabled=""
  groinput=""
/>
`;

exports[`InputDirective Disabled input when focus then check dont have focused class 1`] = `
<input
  class="gro-input gro-disabled"
  disabled=""
  groinput=""
/>
`;

exports[`InputDirective ReactiveForm input check class on focus 1`] = `
<input
  class="gro-input ng-pristine ng-valid ng-touched gro-focused"
  groinput=""
/>
`;

exports[`InputDirective ReactiveForm input should create with formControl 1`] = `
<input
  class="gro-input ng-untouched ng-pristine ng-valid"
  groinput=""
/>
`;

exports[`InputDirective ReactiveForm input should set value from formControl 1`] = `
<input
  class="gro-input ng-untouched ng-pristine ng-valid"
  groinput=""
  value="test"
/>
`;

exports[`InputDirective ReactiveForm input should set value to formControl 1`] = `
<input
  class="gro-input ng-valid ng-touched ng-dirty"
  groinput=""
  value="toto"
/>
`;

exports[`InputDirective ReactiveForm input when control is disabled then check 1`] = `
<input
  class="gro-input ng-untouched ng-pristine gro-disabled"
  disabled=""
  groinput=""
/>
`;

exports[`InputDirective Readonly input do not emit change 1`] = `
<input
  class="gro-input gro-readonly"
  groinput=""
  readonly=""
/>
`;

exports[`InputDirective Readonly input when focus then check dont have focused class 1`] = `
<input
  class="gro-input gro-readonly"
  groinput=""
  readonly=""
/>
`;
